src/wms/src/main/java/org/geoserver/wms/legendgraphic/JSONLegendGraphicBuilder.java

```diff
@@ -0,0 +1,843 @@
+        if (size2 != null) {
+            ret.element(SIZE, printExpression(size2, Number.class));
+        }
+        Expression opacity = graphic.getOpacity();
+        if (opacity != null) {
+            ret.element(OPACITY, printExpression(opacity, Number.class));
+        }
+        Expression rotation = graphic.getRotation();
+        if (rotation != null) {
+            ret.element(ROTATION, printExpression(rotation, Number.class));
+        }
+        ret.element(GRAPHICS, jGraphics);
+        return ret;
+    }
+
+    /**
+     * @param g
+     * @return
+     */
+    private JSONObject processGraphicalSymbol(GraphicalSymbol g) {
+        JSONObject jGraphic = new JSONObject();
+
+        if (g instanceof Mark) {
+            Mark m = ((Mark) g);
+            Expression wkn = m.getWellKnownName();
+            if (wkn != null) {
+                jGraphic.element(MARK, printExpression(wkn, String.class));
+            }
+            jGraphic = processFill(jGraphic, m.getFill());
+            jGraphic = processStroke(jGraphic, m.getStroke());
+            ExternalMark em = m.getExternalMark();
+            if (em != null) {
+                OnLineResource or = em.getOnlineResource();
+                if (or != null) {
+                    try {
+                        jGraphic.element(EXTERNAL_GRAPHIC_URL, or.getLinkage().toURL().toString());
+                    } catch (MalformedURLException e) {
+                        e.printStackTrace();
+                    }
+                }
+                Icon icon = em.getInlineContent();
+                if (icon != null) {
+                    jGraphic.element(ICON, icon.toString());
+                }
+                jGraphic.element(FORMAT, em.getFormat());
+            }
+        } else if (g instanceof ExternalGraphic) {
+            ExternalGraphic eg = (ExternalGraphic) g;
+            try {
+                jGraphic.element(
+                        EXTERNAL_GRAPHIC_URL,
+                        eg.getOnlineResource().getLinkage().toURL().toString());
+            } catch (MalformedURLException e) {
+                e.printStackTrace();
+            }
+            jGraphic.element(EXTERNAL_GRAPHIC_TYPE, eg.getFormat());
+        }
+        return jGraphic;
+    }
+
+    /**
+     * @param ret
+     * @param symbolizer
+     * @return
+     */
+    private JSONObject processLineSymbolizer(JSONObject ret, LineSymbolizer symbolizer) {
+        ret = processStroke(ret, symbolizer.getStroke());
+        if (symbolizer.getPerpendicularOffset() != null) {
+            ret.element(
+                    PERPENDICULAR_OFFSET,
+                    printExpression(symbolizer.getPerpendicularOffset(), Number.class));
+        }
+        return ret;
+    }
+
+    /**
+     * @param ret
+     * @param symbolizer
+     * @return
+     */
+    private JSONObject processPointSymbolizer(JSONObject ret, PointSymbolizer symbolizer) {
+        ret = processGraphic(ret, symbolizer.getGraphic());
```

Comment:
I'd encode vendor options for all symbolizers, as they can all have them, and some are rendering worthy (e.g., controlling randomized symbol placement, see https://docs.geoserver.org/stable/en/user/styling/sld/extensions/randomized.html ) 

Note added for point symbolizer, but applies to all other symbolizers.

Reply:
Is there a reason that handling this at the super class won't work? (Line 760)

Reply:
Oh no, I just missed it, sorry! :-)